File: android/core/src/main/kotlin/dev/skyhop/core/Collision.kt
package dev.skyhop.core

object Collision {
    /** Circle vs axis aligned rectangle, using the closest point of the rectangle to the circle center. */
    fun circleIntersectsRect(
        cx: Float, cy: Float, radius: Float,
        left: Float, top: Float, right: Float, bottom: Float,
    ): Boolean {
        val nearestX = cx.coerceIn(left, right)
        val nearestY = cy.coerceIn(top, bottom)
        val dx = cx - nearestX
        val dy = cy - nearestY
        return dx * dx + dy * dy < radius * radius
    }
}

File: ios/SkyhopCore/Sources/SkyhopCore/Collision.swift
import Foundation

public enum Collision {
    /// Circle vs axis aligned rectangle, using the closest point of the rectangle to the circle center.
    public static func circleIntersectsRect(
        cx: Double, cy: Double, radius: Double,
        left: Double, top: Double, right: Double, bottom: Double
    ) -> Bool {
        let nearestX = min(max(cx, left), right)
        let nearestY = min(max(cy, top), bottom)
        let dx = cx - nearestX
        let dy = cy - nearestY
        return dx * dx + dy * dy < radius * radius
    }
}
